File: JavaScript/StudyPlan/LeetCode75/Q238-ProductOfArrayExceptSelf.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const output = [];
    let lMult = 1;
    let rMult = 1;

    for(let i = nums.length - 1; i >= 0; i--) {
        output[i] = rMult;
        rMult *= nums[i];
    }

    for(let j = 0; j < nums.length; j++) {
        output[j] *= lMult;
        lMult *= nums[j];
    }
    return output;
}


const productExceptSelf = (nums) => {
    const prefix = [];
  
    for(let i = 0; i < nums.length; i++) {
        if(i === 0) {
            prefix[i] = 1;
        } else {
            prefix[i] = nums[i - 1] * prefix[i - 1];
        }
    }
  
    const suffix = [];
    for (let i = nums.length - 1; i >= 0; i--) {
        if(i === nums.length - 1) {
            suffix[i] = 1;
        } else {
            suffix[i] = nums[i + 1] * suffix[i + 1];
        }
    }
  
    const result = [];
    for(let i = 0; i < nums.length; i++) {
        result[i] = prefix[i] * suffix[i];
    }
    return result;
  }

File: Python/Leetcode75/Medium/Array/Q238-ProductOfArrayExceptSelf.py
import math

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [0] * n

        for i in range(0, n):
            l_nums = nums[:i]
            r_nums = nums[i+1:]

            if len(l_nums) == 0:
                ans[i] = 1 * math.prod(r_nums)
            elif len(r_nums) == 0:
                ans[i] = 1 * math.prod(l_nums)
            else:
                ans[i] = math.prod(l_nums) * math.prod(r_nums)
        return ans
    
    def productExceptSelfOptimized(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [0] * n

        l_prods = [1] * n
        for i in range(1, n):
            l_prods[i] = l_prods[i - 1] * nums[i - 1]
        
        r_prods = [1] * n
        for i in range(n - 2, -1, -1):
            r_prods[i] = r_prods[i + 1] * nums[i + 1]
        
        for i in range(n):
            ans[i] = l_prods[i] * r_prods[i]
        
        return ans

    def productExceptSelfOptimized2(self, nums: List[int]) -> List[int]:
        prefix = 1
        suffix = 1
        n = len(nums)
        ans = [0] * n

        for i in range(n):
            ans[i] = prefix
            prefix *= nums[i]

        for i in range(n - 1, -1, -1):
            ans[i] *= suffix
            suffix *= nums[i]

        return ans
 